src/Framework/Framework/Hosting/Middlewares/DotvvmRoutingMiddleware.cs

```diff
@@ -53,12 +53,35 @@ private static bool TryParseGooglebotHashbangEscapedFragment(string queryString,
                 url = url.Substring(HostingConstants.SpaUrlIdentifier.Length).Trim('/');
             }
 
-
             // find the route
+            RouteBase? partialMatch = null;
+            IDictionary<string, object?>? partialMatchParameters = null;
+
             foreach (var r in routes)
             {
-                if (r.IsMatch(url, out parameters)) return r;
+                if (r.IsMatch(url, out parameters))
+                {
+                    isPartialMatch = false;
+                    return r;
+                }
+
+                if (partialMatch == null
```

Comment:
I think that we can speed this up by first trying IsMatch on all routes and only if that fails, try the partial matching. You can also add a second list `partialMatchRoutes` filtered only for the `IPartialMatchRoute` type